### PR TITLE
PresRat correction

### PR DESCRIPTION
**Pull request by castelao**
Implementing PresRat for precipitation.

### CODE REVIEW

File: sup3r/bias/bias_transforms.py
Comment (bnb32) [inline sup3r/bias/bias_transforms.py:673]: doc string here. also `np.ndarray` is the correct type hint, I believe.

File: sup3r/bias/qdm.py
Comment (bnb32) [inline sup3r/bias/qdm.py:593]: lets use `TODO:` to track things like this. It's easy to forget about a task mentioned in an inline comment.

File: sup3r/bias/qdm.py
Comment (bnb32) [inline sup3r/bias/qdm.py:700]: I suppose this is an example of the duplicate code you mentioned. We have a lot of blocks with this kind of pattern, in completely independent classes / modules, and I've often wondered if it could be replaced with a custom decorator or something. Not worth doing right now though.

**Review by bnb32 — CHANGES_REQUESTED** [review]
I don't know the details of the presrat implementation but the mechanics here look good to me. Just a couple tweaks. Also, let's add a fwp integration test like with qdm.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:673]: Opps! Good catch! Thanks.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:598]: Good idea. I'll add it, but I'm hoping to clean this before merging.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:705]: Yes, fully agree. It would be nice to combine in one generic procedure. So far, I've been trying to follow the same patterns in the rest of the library.

I'm glad to work on that here or in a later PR. Your call.

File: sup3r/bias/qdm.py
Comment (grantbuster) [inline sup3r/bias/qdm.py:584]: CDF/QDM parameters are not calculated monthly? I see the mean values to calculate K are calculated monthly but not the CDFs or zero rates? Is this still TODO?

File: sup3r/bias/bias_transforms.py
Comment (grantbuster) [inline sup3r/bias/bias_transforms.py:762]: If we're running GCM future years, shouldn't the zero rate be set using the zero rate from the GCM data? I think the code as written here would fix the zero rate using the historical values. 

Based on this quote from the Pierce paper:
"The GCM predicted future fraction of zero-precipitation days Zgf is calculated using t with the GCM’s original (not bias corrected) future time series."

File: sup3r/bias/bias_transforms.py
Comment (grantbuster) [inline sup3r/bias/bias_transforms.py:769]: Can we really apply the k correction factor given only a few days of modeled future data (`tmp[:, :, idx]` here)? I would think we would need to calculate k monthly but as a multi-year factor based on the extended modeled future time period? 

Also i'm not convinced that k simplifies how you have it here, but this might require a whiteboard to discuss.

File: sup3r/bias/qdm.py
Comment (grantbuster) [inline sup3r/bias/qdm.py:611]: Dont we also need the mean value of the modeled future data after bias correction to calculate x^?

File: tests/bias/test_presrat_bias_correction.py
Comment (grantbuster) [inline tests/bias/test_presrat_bias_correction.py:19]: I think we need to add a test that looks at the fundamental benefits of presrat: 1) preserve the model predicted change in fraction of nonprecipitating days (with the exception of models with strong dry bias) 2) preservation of mean precip change. We can do this with the `fp_fut_cc` file constructors if they're ready or make dummy distributions for single pixels/months in the test.

**Review by grantbuster — CHANGES_REQUESTED** [review]

File: sup3r/bias/qdm.py
Comment (grantbuster) [inline sup3r/bias/qdm.py:597]: We should add an option to calculate the zero rate threshold from the smallest non-zero observed historical value.

File: sup3r/bias/qdm.py
Comment (bnb32) [reply sup3r/bias/qdm.py:710]: Missed the response here. Well put this on the todo list after the refactor integration.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:581]: Fixed/improved.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:610]: Updated the documentation.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:602]: I assumed it was a good first approximation, but based on our last talk. Since that would be required to for the actually forward runs, I already implemented the windowed QDM in #226 , and is already included here. But it would be cleaner and easier if we merge #226 first.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:762]: I think we cleared this one already? If you agree, please close (resolve) this conversation.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:769]: We have a completely different implementation now. I think we should close this one, and if needed open another one with the current implementation.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:721]: We are doing something else now. I'll close this one. Please feel free to reopen if you still want to discuss more on this instead of the new one.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:689]: This is very different to what we do now. I'll also close this one to avoid confusion, but feel free to reopen if you would like to discuss more about this.

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py:822]: Issue #227 so we don't forget about it.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:583]: Fixed.

**Comment by castelao** [comment]
To keep everyone in the same page. This PR is still missing a couple of tests and a cleanup. I'll remove the draft state when it's ready.

**Comment by castelao** [comment]
A reminder for myself. A small edition in the documentation, as suggested by @grantbuster:
"Okay i think I see your point about how this will actually totally work with hourly data because DOY is being used in the indexing. Maybe note in the _window_center docstring that the output array indexes the day of year? Looks good to me! I'm not worried about fractional days for leap years."

**Comment by castelao** [comment]
Just to inform. This PR was rebased on main with the new time dependent QDM, which simplifies this PR since the QDM component was already reviewed thus not part of this PresRat PR.

I have to make some adjustments here to conform with the last changes in QDM, and once it's ready I'll remove the draft state and ping reviewers.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:606]: It's possible if it doesn't require a precise solution. Your call.

What makes this case tricky is to apply a different quantile per gridpoint (the two dimensions in the loop), and the fact that we have more data with low values.

Comment (grantbuster) [reply sup3r/bias/bias_transforms.py:606]: Aren't we setting zero precip rates with tau_fut threshold values now? There is 100% an easy way to do this with broadcasted code. How about we take the raster of tau_fut data, repeat/reshape it to match the time dimension in the biased data, and then do a `np.where(arr < tau_fut, 0, arr)` operation?

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:584]: I missed this comment before. Fixed

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py:175]: Fixed

File: sup3r/bias/qdm.py
Comment (castelao) [reply sup3r/bias/qdm.py]: All done. I'm closing this one.

File: sup3r/bias/bias_transforms.py
Comment (castelao) [reply sup3r/bias/bias_transforms.py]: A note for myself. Remove the apply_zero_rate and related tests.

Comment (castelao) [reply sup3r/bias/bias_transforms.py]: Removed.

File: tests/bias/test_presrat_bias_correction.py
Comment (castelao) [reply tests/bias/test_presrat_bias_correction.py:21]: On my notes, this is the last requirement.

File: sup3r/bias/bias_transforms.py
Comment (grantbuster) [reply sup3r/bias/bias_transforms.py]: pointing to QDM is correct, ignore this comment.

File: sup3r/bias/bias_transforms.py
Comment (grantbuster) [reply sup3r/bias/bias_transforms.py]: @castelao  i think this comment is outstanding

**Comment by grantbuster** [comment]
@castelao, i took another look and the implementation looks great. I know you're adding a presrat-motivation test (e.g., testing the reason why we're doing presrat) which should be great. Could you also clean up docstrings in `bias_transforms.py` and inline comments in `qdm.py`? Specifically, there are just some docstrings in `bias_transforms.py` that have arguments out-of-order or are missing returns statements and lots of methods/todo notes in `qdm.py` that should be cleaned up. 

There is also one outstanding comment on `assert data.ndim` in `bias_transforms.py`

**Review by grantbuster — APPROVED** [review]

**Review by grantbuster — APPROVED** [review]

**Review by bnb32 — APPROVED** [review]

**Pull request merged**
